File: scripts/map/parser.py
# Script to read in data about the network.
from classes import Station, Line
from naptan import NaptanReader
from typing import Optional
import util

# Lines of input file that represent constraints, with their network line and input line number (processed later)
constraint_text_lines: list[tuple[str, str, int]] = list()  # todo: create type
lines = dict()                                # line name -> Line object mapping
naptan_reader: Optional[NaptanReader] = None  # Reader for getting naptans
input_line_number: int = 0                    # Current line of the input being processed, for error messages


def sanitize_line(line: str) -> str:
    """
    :param line: Line of input.
    :return: Line with comments removed, and stripped of leading/trailing whitespace.
    """
    try:
        comment_index = line.index("#")
        line = line[:comment_index]
    except ValueError:
        pass
    line = line.strip()
    return line


def create_new_line(input_file_line: str) -> str:
    """
    Creates a new line in the network. Also returns the name of the line created.
    :param input_file_line: line of input which declares a line.
    :return: The name of the new line.
    """
    tokens = input_file_line.split()
    try:
        line_name = tokens[1]
    except IndexError:
        raise ValueError(f"line {input_line_number} '{input_file_line}' does not have a line name.")

    try:
        colon_index = line_name.index(":")
        line_name = line_name[:colon_index]
    except ValueError:
        pass
    line_name = line_name.strip()

    if line_name in lines:
        raise ValueError(f"redefinition of line '{line_name}'.")

    lines[line_name] = Line(line_name)
    return line_name


def create_new_station(input_file_line: str, current_line: str) -> None:
    """
    Creates a new station.
    :param input_file_line: line of input which declares a station.
    :param current_line: the current line that is selected in the network.
    """
    if current_line is None:
        raise ValueError(f"line {input_line_number} '{input_file_line}' was declared before a current line was set.")
    old_input_file_line = input_file_line

    input_file_line.removeprefix("station")
    input_file_line.lstrip()

    try:
        station_name, rest = util.consume_double_quoted(input_file_line)
        station_name.strip()
        rest.strip()
    except ValueError:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' does not have a double-quoted station name.")

    tokens = rest.split()
    try:
        x = int(tokens[0])
        y = int(tokens[1])
    except IndexError:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' does not have both coordinates.")
    except ValueError:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' contains coordinate which is not an integer.")

    naptan = naptan_reader.get_naptan(station_name)
    station = Station(current_line, station_name, naptan, x, y)
    lines[current_line].add_station(station)


def add_edges(input_file_line: str, current_line: str) -> None:
    """
    Adds a number of edges between two stations on a line.
    :param input_file_line: line of input which declares a series of edges.
    :param current_line: the current line that is selected in the network.
    """
    if current_line is None:
        raise ValueError(f"line {input_line_number} '{input_file_line}' was declared before a current line was set.")
    old_input_file_line = input_file_line

    input_file_line.removeprefix("edges")
    input_file_line.lstrip()

    try:
        stations_string, __ = util.consume_double_quoted(input_file_line)
        stations_string.strip()
    except ValueError:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' does not have double-quoted stations.")

    stations = stations_string.split(",")
    stations = [station.strip() for station in stations]
    if len(stations) < 2:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' has fewer than two stations.")

    current_station = stations[0]
    index = 1
    while index < len(stations):
        next_station = stations[index]
        lines[current_line].add_edge(current_station, next_station, input_line_number)
        current_station = next_station
        index += 1


def validate_curve_type(rest: str, old_input_file_line: str) -> str:
    """
    Extracts the curve type from a line of input and validates it.
    :param rest: the rest of the line that declared the curve, after 'curve'
    :param old_input_file_line: line of input which declared the curve, for error messages.
    :return: The extracted curve type.
    """
    rest_tokens = rest.split()
    curve_type = rest_tokens[0]
    curve_type_tokens = curve_type.split(",")
    if curve_type != "special":
        error = False
        if len(curve_type_tokens) != 2:
            error = True
        for token in curve_type_tokens:
            token = token.strip()
            if token not in ["up", "down", "left", "right", "up-right", "up-left", "down-right", "down-left"]:
                error = True
        if error:
            raise ValueError(f"line {input_line_number} '{old_input_file_line}' has invalid curve type.")

    return curve_type


def add_curve(input_file_line: str, current_line: str) -> None:
    """
    :param input_file_line: line of input which declares a curve.
    :param current_line: the current line that is selected in the network.
    """
    if current_line is None:
        raise ValueError(f"line {input_line_number} '{input_file_line}' was declared before a current line was set.")
    old_input_file_line = input_file_line

    input_file_line.removeprefix("curve")
    input_file_line.lstrip()

    try:
        stations_string, rest = util.consume_double_quoted(input_file_line)
        stations_string.strip()
        rest.strip()
    except ValueError:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' does not have double-quoted stations.")
    if len(rest) == 0:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' does not have curve type.")

    curve_type = validate_curve_type(rest, old_input_file_line)

    stations = stations_string.split(",")
    stations = [station.strip() for station in stations]
    if len(stations) != 2:
        raise ValueError(f"line {input_line_number} '{old_input_file_line}' does not have two stations.")

    lines[current_line].add_curve(stations[0], stations[1], curve_type, input_line_number)


def read_data(path) -> None:
    """
    Reads data about lines/stations from the input file. Puts line information into the ``lines`` global variable,
    and leaves constraints unprocessed in the ``constraint_text_lines`` global variable.
    :param path: Path to the input file.
    """
    global input_line_number
    station_data_file = open(path, "r")

    current_line = None
    for input_file_line in station_data_file.readlines():
        input_line_number += 1
        input_file_line = sanitize_line(input_file_line)
        if len(input_file_line) == 0:
            continue

        if input_file_line.startswith("line"):
            current_line = create_new_line(input_file_line)
        elif input_file_line.startswith("station"):
            create_new_station(input_file_line, current_line)
        elif input_file_line.startswith("edges"):
            add_edges(input_file_line, current_line)
        elif input_file_line.startswith("curve"):
            add_curve(input_file_line, current_line)
        elif input_file_line.startswith("multi-line"):
            current_line = "_MULTI"
        elif input_file_line.startswith(tuple(["vertical", "horizontal", "up-right",
                                               "up-left", "down-right", "down-left"])):
            input_file_line = input_file_line.replace("up-left", "down-right")
            input_file_line = input_file_line.replace("down-left", "up-right")
            constraint_text_lines.append((input_file_line, current_line, input_line_number))
        elif input_file_line.startswith("same-station") or input_file_line.startswith("equal"):
            constraint_text_lines.append((input_file_line, current_line, input_line_number))
        else:
            raise ValueError(f"line {input_line_number} '{input_file_line}' has unrecognized form.")


def check_no_orphans() -> None:
    """Checks that all lines have no orphans."""
    for line in lines.values():
        has_orphan, orphan = line.has_orphan_stations()
        if has_orphan:
            raise ValueError(f"line {line.name} has orphan station {orphan}.")


def parse_data() -> tuple[dict, list[tuple[str, str, int]]]:
    global naptan_reader
    naptan_reader = NaptanReader("input/naptan.json")
    read_data("input/tube_data.txt")
    check_no_orphans()
    return lines, constraint_text_lines


if __name__ == "__main__":
    print("This file is not mean to be run. Run generate_map.py instead.")
    exit(1)

File: scripts/map/src/main/java/com/github/alexandergillon/mini_metro_maps/Parser.java
package com.github.alexandergillon.mini_metro_maps;

import com.github.alexandergillon.mini_metro_maps.models.core.AlignmentConstraint;
import com.github.alexandergillon.mini_metro_maps.models.core.Curve;
import com.github.alexandergillon.mini_metro_maps.models.core.MetroLine;
import com.github.alexandergillon.mini_metro_maps.models.core.Station;
import com.github.alexandergillon.mini_metro_maps.models.core.ZIndexConstraint;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.commons.lang3.tuple.Triple;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Class to read in data about the metro network. */
public class Parser {

    /** Mapping from metro line name -> MetroLine object. */
    private final HashMap<String, MetroLine> metroLines = new HashMap<>();

    /** Constraints declared in the input file. */
    private final ArrayList<AlignmentConstraint> alignmentConstraints = new ArrayList<>();

    /** Path to the input data file. */
    private final String inputPath;

    /** For getting NAPTANs of stations. */
    private final NaptanReader naptanReader;

    /** Current metro line active in the input file. */
    MetroLine currentMetroLine = null;

    /** Current line of the input being processed, for error messages. */
    private int textLineNumber = 0;

    /** z-index constraints among lines. */
    private final Set<ZIndexConstraint> zIndexConstraints = new HashSet<>();

    /**
     * List of curves which are parallel to other curves, the text which describes who they are parallel to, and the
     * line number that declared them.
     */
    private final ArrayList<Triple<Curve, String, Integer>> parallelCurves = new ArrayList<>();

    /**
     * @param inputPath Path to the input data file.
     * @param naptanFilepath Path to naptan.json.
     */
    public Parser(String inputPath, String naptanFilepath) throws IOException {
        this.inputPath = inputPath;
        naptanReader = new NaptanReader(naptanFilepath);
    }

    /**
     * Throws a formatted IllegalArgumentException. This exists to make many instances of
     * IllegalArgumentException(String.format(...)) more concise.
     */
    private void parseException(String format, Object... args) {
        throw new IllegalArgumentException(String.format(format, args));
    }

    /**
     * @param textLine A line of text input.
     * @return That line with comments removed, and stripped of leading/trailing whitespace.
     */
    private String sanitizeTextLine(String textLine) {
        int commentIndex = textLine.indexOf('#');
        if (commentIndex != -1) {
            textLine = textLine.substring(0, commentIndex);
        }
        textLine = textLine.strip();
        return textLine;
    }

    /**
     * Creates a new metro line in the network. Also returns the metro line that was created.
     * @param textLine line of text input which declares a metro line.
     * @return The new metro line object, which has been added to metroLines.
     */
    private MetroLine createNewMetroLine(String textLine) {
        String[] tokens = textLine.split("\\s+"); // splits on whitespace
        if (tokens.length == 0) parseException("Line %d does not have a metro line name.", textLineNumber);

        String metroLineName = tokens[1];
        int colonIndex = metroLineName.indexOf(':');
        if (colonIndex != -1) {
            metroLineName = metroLineName.substring(0, colonIndex);
        }
        metroLineName = metroLineName.strip();

        if (metroLines.containsKey(metroLineName)) parseException("Redefinition of line %s.", metroLineName);

        MetroLine metroLine = new MetroLine(metroLineName);
        metroLines.put(metroLineName, metroLine);
        return metroLine;
    }

    /**
     * Creates a new station.
     * @param textLine Line of input text which declares a station.
     * @param currentMetroLine The current metro line that is selected in the network.
     * @param isAlignmentPoint Whether this station is actually a station, or just an alignment point for special edges.
     */
    private void createNewStation(String textLine, MetroLine currentMetroLine, boolean isAlignmentPoint) {
        if (currentMetroLine == null) parseException("(line %d) Station declared before a current line was set.", textLineNumber);

        textLine = isAlignmentPoint ? Util.removePrefix(textLine, "alignment-point")
                : Util.removePrefix(textLine, "station").strip();

        Pair<String, String> doubleQuotedResult = Util.consumeDoubleQuoted(textLine, textLineNumber);
        String stationName = doubleQuotedResult.getLeft().strip();
        String textRest = doubleQuotedResult.getRight().strip();

        String[] tokens = textRest.split("\\s+");
        if (tokens.length != 2) parseException("(line %d) Station declaration does not have both coordinates, or has trailing text.", textLineNumber);

        int x, y;
        try {
            x = Integer.parseInt(tokens[0]);
            y = Integer.parseInt(tokens[1]);
        } catch (NumberFormatException ignored) {
            // Can't use parseException or compiler cannot prove that x and y are initialized.
            throw new IllegalArgumentException(String.format("(line %d) Station declaration contains coordinate which is not an integer.", textLineNumber));
        }

        // todo: maybe move ampl id out here
        String naptan = isAlignmentPoint ? "alignmentpoint_" + stationName
                : naptanReader.getNaptan(currentMetroLine.getName(), stationName);
        Station station = new Station(currentMetroLine.getName(), stationName, naptan, x, y, isAlignmentPoint);
        currentMetroLine.addStation(station);
    }

    /**
     * Adds a number of edges between two stations on a line.
     * @param textLine Line of input text which declares a number of edges.
     * @param currentMetroLine The current metro line that is selected in the network.
     */
    private void addEdges(String textLine, MetroLine currentMetroLine) {
        if (currentMetroLine == null) parseException("(line %d) Edges declared before a current line was set.", textLineNumber);

        textLine = Util.removePrefix(textLine, "edges").strip();

        Pair<String, String> doubleQuotedResult = Util.consumeDoubleQuoted(textLine, textLineNumber);
        String stationsString = doubleQuotedResult.getLeft().strip();

        List<Pair<String, String>> stationPairs = Util.allConsecutiveStationPairs(stationsString, textLineNumber);
        for (Pair<String, String> stationPair : stationPairs) {
            currentMetroLine.addEdge(stationPair.getLeft(), stationPair.getRight(), textLineNumber);
        }
    }

    /**
     * Validates a curve type.
     * @param curveType The curve type to validate.
     */
    private void validateCurveType(String curveType) {
        if (!curveType.equals("special")) {
            String[] curveTypeTokens = curveType.split(",");
            if (curveTypeTokens.length != 2) parseException("(line %d) Invalid curve type \"%s\".", textLineNumber, curveType);

            String[] validTokens = {"up", "down", "left", "right", "up-right", "up-left", "down-right", "down-left"};
            if (!ArrayUtils.contains(validTokens, curveTypeTokens[0]) || !ArrayUtils.contains(validTokens, curveTypeTokens[1])) {
                parseException("(line %d) Invalid curve type \"%s\".", textLineNumber, curveType);
            }
        }
    }

    /**
     * Process a 'parallelto' clause in a curve declaration. For now, adds it to the map of parallel curves - these
     * dependencies can only be resolved when all curves have been read in.
     * @param inputText Text that represents a parallelto clause.
     * @param curve The curve that was declared with this 'parallelto' clause.
     */
    private void processParallelTo(String inputText, Curve curve) {
        Pair<String, String> keywordAndRest = Util.consumeToken(inputText, textLineNumber);
        String keyword = keywordAndRest.getLeft();
        String textRest = keywordAndRest.getRight();

        if (!keyword.equals("parallelto")) parseException("(line %d) Unrecognized trailing text on curve declaration.", textLineNumber);

        parallelCurves.add(Triple.of(curve, textRest, textLineNumber));
    }

    /**
     * Adds a curve between two stations.
     * @param textLine Line of input text which declares a curve.
     * @param currentMetroLine The current metro line that is selected in the network.
     */
    private void addCurve(String textLine, MetroLine currentMetroLine) {
        if (currentMetroLine == null) parseException("(line %d) Curve declared before a current line was set.", textLineNumber);

        textLine = Util.removePrefix(textLine, "curve").strip();

        Pair<String, String> doubleQuotedResult = Util.consumeDoubleQuoted(textLine, textLineNumber);
        String stationsString = doubleQuotedResult.getLeft().strip();
        String textRest = doubleQuotedResult.getRight().strip();

        Pair<String, String> curveTypeAndRest = Util.consumeToken(textRest, textLineNumber);
        String curveType = curveTypeAndRest.getLeft();
        textRest = curveTypeAndRest.getRight();

        String[] stations = Arrays.stream(stationsString.split(",")).map(String::strip).toArray(String[]::new);
        validateCurveType(curveType);

        if (stations.length != 2) parseException("(line %d) Curve declaration does not have two stations.", textLineNumber);

        Curve.SpecialCurveInfo specialCurveInfo = curveType.equals("special") ?
                Curve.SpecialCurveInfo.fromText(textRest, currentMetroLine, textLineNumber) : null;
        Curve curve = currentMetroLine.addCurve(stations[0], stations[1], curveType, specialCurveInfo, textLineNumber);

        if (!textRest.isEmpty()) {
            try {
                processParallelTo(textRest, curve);
            } catch (IllegalArgumentException e) {
                // Curve has trailing text which is not 'parallelto'. If the curve is 'special', that is ok.
                if (!curveType.equals("special")) {
                    throw e;
                }
            }
        }
    }

    /**
     * Adds an endpoint to a metro line.
     * @param textLine Line of input text which declares an endpoint.
     * @param currentMetroLine The current metro line that is selected in the network.
     */
    private void addEndpoint(String textLine, MetroLine currentMetroLine) {
        if (currentMetroLine == null) parseException("(line %d) Endpoint declared before a current line was set.", textLineNumber);

        textLine = Util.removePrefix(textLine, "endpoint").strip();

        Pair<String, String> doubleQuotedResult = Util.consumeDoubleQuoted(textLine, textLineNumber);
        String stationString = doubleQuotedResult.getLeft().strip();
        String textRest = doubleQuotedResult.getRight().strip();

        currentMetroLine.addEndpoint(stationString, textRest, textLineNumber);
    }

    /**
     * Adds a z-index constraint between two lines.
     * @param textLine Line of input text which declares a z-index constraint.
     */
    private void addZIndexConstraint(String textLine) {
        textLine = Util.removePrefix(textLine, "zindex").strip();
        String[] tokens = textLine.split("\\s+");

        if (tokens.length != 3) parseException("(line %d) z-index constraint declaration is invalid.", textLineNumber);

        String metroLineName1 = tokens[0];
        String constraintType = tokens[1];
        String metroLineName2 = tokens[2];

        switch (constraintType) {
            case "above" -> zIndexConstraints.add(new ZIndexConstraint(metroLineName1, metroLineName2));
            case "below" -> zIndexConstraints.add(new ZIndexConstraint(metroLineName2, metroLineName1));
            default -> parseException("(line %d) Bad z-index constraint type \"%s\".", textLineNumber, constraintType);
        }
    }

    /**
     * Processes a sanitized line of input text.
     * @param textLine The line of input text to process.
     */
    private void processInputLine(String textLine) {
        if (textLine.startsWith("line")) {
            currentMetroLine = createNewMetroLine(textLine);
        } else if (textLine.startsWith("station")) {
            createNewStation(textLine, currentMetroLine, false);
        } else if (textLine.startsWith("alignment-point")) {
            createNewStation(textLine, currentMetroLine, true);
        } else if (textLine.startsWith("edges")) {
            addEdges(textLine, currentMetroLine);
        } else if (textLine.startsWith("curve")) {
            addCurve(textLine, currentMetroLine);
        } else if (textLine.startsWith("multi-line")) {
            currentMetroLine = null;
        } else if (Util.startsWithAny(textLine, new String[]{"vertical", "horizontal", "up-right", "up-left", "down-right", "down-left"})) {
            textLine = textLine.replace("up-left", "down-right");
            textLine = textLine.replace("down-left", "up-right");
            alignmentConstraints.add(new AlignmentConstraint(textLine, currentMetroLine, textLineNumber));
        } else if (Util.startsWithAny(textLine, new String[]{"same-station", "equal"})) {
            alignmentConstraints.add(new AlignmentConstraint(textLine, currentMetroLine, textLineNumber));
        } else if (textLine.startsWith("endpoint")) {
            addEndpoint(textLine, currentMetroLine);
        } else if (textLine.startsWith("zindex")) {
            addZIndexConstraint(textLine);
        } else {
            System.out.printf("(line %d) Warning: line with unrecognized form. Ignoring.%n", textLineNumber);
        }
    }

    /**
     * Reads data about metro lines/stations from the input file. Puts information into the `metroLines` and
     * `constraints` member variables.
     */
    private void readData() throws IOException {
        try (BufferedReader bufferedReader = new BufferedReader(new FileReader(inputPath))) {
            String textLine;
            textLineNumber = 0;
            while ((textLine = bufferedReader.readLine()) != null) {
                textLineNumber++;
                textLine = sanitizeTextLine(textLine);
                if (textLine.isEmpty()) continue;

                processInputLine(textLine);
            }
        }
    }

    /** Checks that all lines have no orphans. */
    private void checkNoOrphans() {
        for (MetroLine metroLine : metroLines.values()) {
            metroLine.assertNoOrphanStations();
        }
    }

    /** Resolves a curve specifier (e.g. "picadilly: Northfields, South Ealing" to a curve object. */
    private Curve resolveCurve(String curveSpecifier) {
        Pair<String, String> doubleQuotedResult = Util.consumeDoubleQuoted(curveSpecifier, textLineNumber);

        String[] tokens = doubleQuotedResult.getLeft().split(":");
        if (tokens.length != 2) parseException("(line %d) Unrecognized trailing text on curve declaration.", textLineNumber);

        String[] stations = tokens[1].split(",");
        if (stations.length != 2) parseException("(line %d) Unrecognized trailing text on curve declaration.", textLineNumber);

        String metroLineName = tokens[0].strip();
        String station1 = stations[0].strip();
        String station2 = stations[1].strip();

        if (!metroLines.containsKey(metroLineName)) parseException("(line %d) Unrecognized trailing text on curve declaration.", textLineNumber);

        return metroLines.get(metroLineName).getCurve(station1, station2, textLineNumber);
    }

    /**
     * Resolves a curve with the same 'from' and 'to' stations on a different line.
     * @param curve A curve.
     * @param metroLineName A line name.
     * @return A curve on that line, with the same 'from' and 'to' stations as the curve.
     */
    private Curve resolveCurve(Curve curve, String metroLineName) {
        if (!metroLines.containsKey(metroLineName)) parseException("(line %d) Unrecognized trailing text on curve declaration.", textLineNumber);

        Curve otherCurve = metroLines.get(metroLineName).getCurve(curve.getFrom().getName(), curve.getTo().getName(), textLineNumber);

        if (!otherCurve.getType().equals(curve.getType())) {
            parseException("(line %d) Curve with type %s declared to be dependent on curve with type %s.",
                    textLineNumber, curve.getType(), otherCurve.getType());
        }

        return otherCurve;
    }

    /** Resolves all curve dependencies (sets the Curve.parallelTo field for all curves appropriately). */
    private void resolveCurveDependencies() {
        for (Triple<Curve, String, Integer> dependency : parallelCurves) {
            Curve curve = dependency.getLeft();
            String parallelToText = dependency.getMiddle();
            textLineNumber = dependency.getRight();

            Pair<String, String> firstTokenAndRest = Util.consumeToken(parallelToText, textLineNumber);
            String firstToken = firstTokenAndRest.getLeft();
            String textRest = firstTokenAndRest.getRight().strip();

            // If the first token is 'curve' - read the curve from a double-quoted string. Otherwise, the first
            // token is the line name of a line that has the exact same curve as this (between same stations, same
            // direction, etc.).
            Curve parallelTo = firstToken.equals("curve") ? resolveCurve(textRest) : resolveCurve(curve, firstToken);
            curve.setParallelTo(parallelTo);
        }
    }

    /**
     * Parses data from the input file.
     * @return A triple of the following:
     *   - A mapping from metro line name -> MetroLine object.
     *   - A list of alignment constraints.
     *   - A list of z-index constraints.
     */
    public Triple<Map<String, MetroLine>, List<AlignmentConstraint>, Set<ZIndexConstraint>> parseData() throws IOException {
        System.out.println("Parsing input data.");
        readData();
        checkNoOrphans();
        resolveCurveDependencies();
        return Triple.of(metroLines, alignmentConstraints, zIndexConstraints);
    }
}
